### PR TITLE
Pipeline ASR with LM.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 1]
```diff
@@ -241,16 +251,22 @@ def _forward(self, model_inputs):
...
-        skip_special_tokens = False if "CTC" in self.tokenizer.__class__.__name__ else True
-        tokens = np.concatenate([outputs["tokens"].numpy() for outputs in model_outputs], axis=-1)
-        tokens = tokens.squeeze(0)
-
-        recognized_string = self.tokenizer.decode(tokens, skip_special_tokens=skip_special_tokens)
-        return {"text": recognized_string}
+        if self.type == "ctc_with_lm":
+            logits = np.concatenate([outputs["logits"].numpy() for outputs in model_outputs], axis=1)
+            logits = logits.squeeze(0)
+            text = self.tokenizer.decode_beams(logits)[0][0]
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/automatic_speech_recognition.py:266]: not supper happy that the decoder is now the tokenizer in terms of naming, but I see why it makes sense! Also should we allow the user to pass **kwargs to `decode_beams(...)`?

**Review by patrickvonplaten — COMMENTED** [review]
looks very nice already!

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:266]: Yes, I hesitated back&forth. Adding a decoder, means adding a bunch of code everywhere to automatically load it up etc.. For a single pretty niche (currently) use case.
Adding it as tokenizer, is defendable as you said, and makes things simpler for all utils. It's not however really a tokenizer (since it takes `logits` not `input_ids`). The good thing, is that for the pipeline it should roughly be seen as an implementation detail, not really something users should depend on  (if it's `tokenizer`).

I don't like adding `kwargs` without careful thought (pipelines are already a kwarg mess, adding on top is something worth at least a pause).

- Isn't there a way to make the decoder itself have the kwargs ? That means better ownership of information. (overridding kwargs means the user now has to declare himself `decoder = pytctc.Decode(..., **kwargs); pipe = pipeline(..., decoder=decoder)` and at least who eats those arguments is extra clear.
- Adding them to the raw pipeline is like the `generate_kwargs`, not every argument can be changed in the future, and user has no idea if the argument was correctly passed. (Misnamed argument gets silently discarded)
- Other option if `decoder` doesn't own it is to have `pipeline(..., decoder_kwargs={..})`. This is cleaner as it adds only 1 kwargs and it's unlikely to be miss attributed. However, those are not "owned" by the decoder itself, meaning the behavior is pipeline specific. Going that route the arguments to pipelines are going to get extremely long (they already are).

Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py]: I moved to `decoder` in the end.

**Comment by Narsil** [comment]
I think it's ready for final review.

Not a lot of changes overall since last time:

- `pipeline(model="...")` works by default, and fallbacks to regular CTC if `pyctcdecode` missing, or any other error.
- the `decoder` is now a real `decoder` (easier for the fallback actually).

File: src/transformers/pipelines/__init__.py [hunk 2]
```diff
@@ -611,6 +611,21 @@ def pipeline(
...
 
+            if (
+                feature_extractor._processor_class
+                and feature_extractor._processor_class.endswith("WithLM")
+                and isinstance(model_name, str)
+            ):
+                try:
+                    from pyctcdecode import BeamSearchDecoderCTC
+
+                    decoder = BeamSearchDecoderCTC.load_from_hf_hub(model_name)
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/__init__.py:622]:
```suggestion
            language_model_filenames os.path.join(BeamSearchDecoderCTC._LANGUAGE_MODEL_SERIALIZED_DIRECTORY, "*")
            alphabet_filename = BeamSearchDecoderCTC._ALPHABET_SERIALIZED_FILENAME
            allow_regex = [language_model_filenames, alphabet_filename]
            
            decoder = BeamSearchDecoderCTC.load_from_hf_hub(
                pretrained_model_name_or_path, allow_regex=allow_regex, **kwargs
            )
```

Comment (patrickvonplaten) [reply src/transformers/pipelines/__init__.py:622]: This will prevent the load command to unnecessarily download model files, etc...

**Review by patrickvonplaten — APPROVED** [review]
Awesome - thanks a lot for working on it. Left one final thing to change - which would be to make sure no unnecessary files are downloaded

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 3]
```diff
@@ -144,7 +144,18 @@ def __init__(self, feature_extractor: Union["SequenceFeatureExtractor", str], *a
             raise ValueError("The AutomaticSpeechRecognitionPipeline is only available in PyTorch.")
 
         self.check_model_type(dict(MODEL_FOR_SPEECH_SEQ_2_SEQ_MAPPING.items() + MODEL_FOR_CTC_MAPPING.items()))
-        self.is_ctc = self.model.__class__ in MODEL_FOR_CTC_MAPPING.values()
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/automatic_speech_recognition.py:147]: @Narsil - I think you accidently removed this no? It's weird that the tests didn't fail as I think the ASR pipelines don't work on master at the moment

Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:147]: No this was replaced by `self.type` ({`ctc`, `ctc_with_lm` , `seq2seq`)
